_inc/search/components/search-widget.jsx

```diff
@@ -0,0 +1,78 @@
+/** @jsx h */
+
+/**
+ * External dependencies
+ */
+import { h, Component } from 'preact';
+import Portal from 'preact-portal';
+import { debounce } from 'lodash';
+
+class SearchWidget extends Component {
+	constructor() {
+		super( ...arguments );
+		this.requestId = 0;
+		this.state = {
+			query: this.props.initialValue,
+			results: [],
+		};
+		this.onChangeQuery = this.onChangeQuery.bind( this );
+		this.getResults = debounce( this.getResults.bind( this ), 500 );
+		this.getResults( this.props.initialValue );
+	}
+	componentDidMount() {
+		if ( this.props.grabFocus ) {
+			this.input.focus();
+		}
+	}
+
+	bindInput = input => ( this.input = input );
+	onChangeQuery = event => {
+		const query = event.target.value;
+		this.setState( { query } );
+		this.getResults( query );
+	};
+
+	getResults( query ) {
+		if ( query ) {
+			const { api } = this.props;
+			if ( api ) {
+				this.requestId++;
+				const requestId = this.requestId;
+				api
+					.fetch( query )
+					.then( response => {
+						return response.json();
+					} )
+					.then( json => {
+						if ( this.requestId === requestId ) {
+							this.setState( { results: json } );
+						}
+					} );
+			}
+		} else {
+			this.setState( { results: [] } );
+		}
+	}
```

Comment:
I'd strongly recommend replacing `JetpackSearchAPI` and the request handling in this component by a `@wordpress/data`-style store, which can take care of most of these things for us (by virtue of the concept of resolvers in particular).

See the example at https://github.com/WordPress/gutenberg/blob/c5f3c445ae858ba38a23d3f2981bcc8258565a9a/packages/data/README.md#registering-a-store -- it's similar enough to what we're trying to do here that you can essentially copy-paste it with a few modifications. LMK if you want me to help with that.

(Another -- real-live -- example is here: https://github.com/Automattic/jetpack/pull/12823/files#diff-1e7fffc443acdbb76875dedea0598f3a)

Reply:
Yes resolvers take care of a lot of problems, but the tradeoff is bundle size. As this is a front end widget that is potentially loaded for every visitor, we wanted to keep the dependencies as light as possible. I'll take a look, but as I remember the data package pulls in quite a lot of stuff.

Reply:
Thanks! Bummer about the data package pulling in lots of stuff :slightly_frowning_face: TBH, I'm always worried that a workaround pattern like this will proliferate through copy-paste to places where it isn't strictly needed (e.g. editor view), when we have an established pattern.

It's really unfortunate that we find ourselves working around the tooling that Core (the `@wordpress/` packages, specifically) provides us with, out of concern for bundle size :slightly_frowning_face: This seems very unwieldy; I'd personally rather use that tooling, and try to address its size upstream.

Reply:
Yeah, unfortunately the base react bundle is 113k (uncompressed) which is before we bring in any of the `@wordpress` packages. That's quite a hit for a widget that's potentially going on every page for every user.

(also, this is exactly the type of discussion I wanted to figure out here, I know the widget itself needs work and enhancements ;) but the actual inclusion mechanism is my worry)